### ISSUE TITLE
MonitorDownload: doesn't call endFunction on a welcome page

### ISSUE THREAD

**Issue by VsevolodGolovanov** [primefaces/primefaces#3108]
## 1) Environment
- PrimeFaces version: 5.2.12.
- Does it work on the newest released PrimeFaces version? No.
- Does it work on the newest sources in GitHub? 6.2-SNAPSHOT: no.
- Application server + version: Wildfly 11.
- Affected browsers: Firefox 57.0.2. Didn't test others.

## 2) Expected behavior

monitorDownload calling endFunction when the file download is over.

## 3) Actual behavior

monitorDownload never calls endFunction.

## 4) Steps to reproduce

Open the xhtml by a direct URL: http://127.0.0.1:8080/reproducer/view/index.xhtml
Click "Download". Observe the "Downloading..." dialog open and close, when the download is finished.

Now open it via context path URL: http://127.0.0.1:8080/reproducer/
Click "Download". Observe the "Downloading..." dialog open and stay open forever.

## 5) Sample XHTML

	<h:html xmlns="http://www.w3.org/1999/xhtml"
		xmlns:ui="http://java.sun.com/jsf/facelets"
		xmlns:h="http://java.sun.com/jsf/html"
		xmlns:f="http://java.sun.com/jsf/core"
		xmlns:p="http://primefaces.org/ui">
	<f:view locale="en">
	<h:head>
	</h:head>
	<h:body>
		<script>
			function showStatus() {
				PF('statusDialog').show();
			}
			function hideStatus() {
				PF('statusDialog').hide();
			}
		</script>
		<h:form>
			<p:dialog modal="true" widgetVar="statusDialog" header="Status" draggable="false" closable="false">
				Downloading...
			</p:dialog>
			<p:commandButton value="Download" ajax="false" 
				onclick="PrimeFaces.monitorDownload(showStatus, hideStatus)">
				<p:fileDownload value="#{fileDownloadBean.file}"/>
			</p:commandButton>
		</h:form>
	</h:body>
	</f:view>
	</h:html>

## 6) Sample bean

	import java.io.ByteArrayInputStream;
	import java.io.InputStream;
	import java.io.Serializable;
	import java.util.Random;
	import javax.faces.view.ViewScoped;
	import javax.inject.Named;
	import org.primefaces.model.StreamedContent;
	import org.primefaces.model.DefaultStreamedContent;

	@Named
	@ViewScoped
	public class FileDownloadBean implements Serializable {

		private static final long serialVersionUID = 1L;

		private StreamedContent file;

		public FileDownloadBean() {
			byte[] b = new byte[100000000]; // ~100MB
			new Random().nextBytes(b);
			InputStream stream = new ByteArrayInputStream(b); 
			file = new DefaultStreamedContent(stream, "application/octet-stream", "downloaded_file.bin");
		}

		public StreamedContent getFile() {
			return file;
		}

	}

## 7) web.xml

	<welcome-file-list>
		<welcome-file>view/index.xhtml</welcome-file>
	</welcome-file-list>

index.xhtml MUST be in a subfolder, not in context path root.

## 8) Notes.

FileDownloadActionListener sees the rewritten URL, with the `view/` part, and it adds the cookie to this path. But on the client it searches the cookie according to the requested location, the context path.

![dwnldcookie](https://user-images.githubusercontent.com/8070317/34327887-68fd3402-e8e1-11e7-93df-1c415122034f.png)

What could be the fix? Always storing the cookies on the context path, add viewIds to the cookie names instead (that's not quite the same of course - one view can be accessible on multiple URLs). Unless there is a portable way to figure out an actual URL, before it's rewritten?

**Comment by tandraschko** [comment]
a PR is welcome. I think the easiest way is to set the viewId as cookie path in monitorDownload and fileDownload.

**Comment by mertsincan** [comment]
I think users can use monitorKey option for this issue. Please see;
```
...
       monitorDownload: function(start, complete, monitorKey) {
            ...
                var cookieName = monitorKey ? 'primefaces.download_' + monitorKey : 'primefaces.download';
                window.downloadMonitor = setInterval(function() {
                    var downloadComplete = PrimeFaces.getCookie(cookieName);

                    if(downloadComplete === 'true') {
                        ...
                    }
                }, 1000);
            }
        },
...
```

**Comment by VsevolodGolovanov** [comment]
Users can work around the bug with a hack that relies on undocumented implementation details? I guess so...

**Comment by tandraschko** [comment]
@mertsincan this is not directly related to the issue
@VsevolodGolovanov  instead of write critism, why not just try to analyse the problem and provide a possible solution / PR?

**Comment by VsevolodGolovanov** [comment]
Reporting issues is contributing too. I've analyzed it quite a bit in the initial report already.

**Comment by tandraschko** [comment]
Sure it is but code contributions are really welcome. Compared to the open issues, we are to few maintainers.

**Comment by mertsincan** [comment]
Hi,

What do you think about the following fix?
```
//FileDownloadActionListener.java
Map<String, Object> cookieProperties = new HashMap<>();
cookieProperties.put("path", "/"); // Always add cookies to root path
externalContext.addResponseCookie(Constants.DOWNLOAD_COOKIE + monitorKeyValue, "true", cookieProperties);
```

```
//core.js
PrimeFaces.setCookie(cookieName, null, { path: '/' }); // to clear cookie
```

If we use viewId, it is the same for http://127.0.0.1:8080/reproducer/view/index.xhtml and http://127.0.0.1:8080/reproducer/ -> /reproducer/view
But when you want to get cookie value using $.cookie('...') on http://127.0.0.1:8080/reproducer/, 
you will not have access to it.

**Comment by tandraschko** [comment]
yep, root path would fix it but we should at least use the context-path here to not share it with other webapps
Another problem would be if you use the monitor in multiple pages
so maybe we should generate a "monitorKey" per default somehow

**Comment by tandraschko** [comment]
Both FileDownload and MonitorUpload uses the following cookie pattern now:

primefaces.download + viewId

and

the contextPath is the cookiePath now

**Comment by elab** [comment]
I'm facing the same problem (end function of monitorDownload is not called) with **dataExporter**.
It looks like the [DataExporter is still setting the cookie named "primefaces.download"](https://github.com/primefaces/primefaces/blob/30596b9758e8db0c5f862e60d7532f68e87e2d9c/primefaces/src/main/java/org/primefaces/component/export/DataExporter.java#L222-L224).

Could you please look into?

PrimeFaces 10.0.2.

**Comment by melloware** [comment]
@elab I am seeing it generate the full cookie + VIEW ID.

![image](https://user-images.githubusercontent.com/4399574/122104392-9e2ebc80-cde5-11eb-9477-e5ddd0f804a0.png)

**Comment by melloware** [comment]
Oh sorry you are talking about DataExporter not File Download.  Can you please open a new ticket with the exact issue you are having?  Please provide a PrimeFaces Test reproducer as well.

**Comment by melloware** [comment]
@elab I just tested it and it works fine for me with DataExporter.  So I have a feeling you have something else wrong in your app.

```xml
<p:commandButton value="CSV" styleClass="p-mr-2 p-mb-2" onclick="PrimeFaces.monitorDownload(showStatus, hideStatus)">
      <p:dataExporter type="csv" target="tbl" fileName="products" />
</p:commandButton>
```

**Comment by elab** [comment]
Thank you, @melloware, I found the difference: In my p:commandButton / p:commandLink there was the `ajax="false"` attribute. Removing it solves the problem :)

(The project is migrating from PF 8, where it was working, to PF 10.)

----
But why does the `ajax` attribute have impact? The documentation is not quite understandable to me.

The [Migration Guide 8.0 -> 10.0/DataExporter](https://primefaces.github.io/primefaces/10_0_0/#/../migrationguide/10_0_0?id=dataexporter) says: 
> DataExporter
> ...
> supports AJAX (so ajax="false" is not required anymore)

But also not prohibited?

&nbsp;

The [10.0 DataExporter documentation](https://primefaces.github.io/primefaces/10_0_0/#/components/dataexporter?id=monitor-status) states:
> Monitor Status
> DataExport is a non-ajax process so ajaxStatus component cannot apply. See FileDownload Monitor Status section to find out how monitor export process. Same solution applies to data export as well.

The `ajax="false"` is mentioned there ^ only once in the example in the section [Custom Export](https://primefaces.github.io/primefaces/10_0_0/#/components/dataexporter?id=custom-export), which probably won't work with the monitorDownload (calling the end function) by the same reason as in my case?

&nbsp;

And in all of the [Show Case DataExporter Examples](https://www.primefaces.org/showcase/ui/data/dataexporter/basic.xhtml) there is no `ajax` attribute at all.

&nbsp;

Interesting that the FileDownload (as opposed to DataExporter) does work (i.e. the end function of monitorDownload gets called) in both cases: with and without `ajax="false"` in the surrounding commandButton/commandLink.

PrimeFaces 10.0.2

**Comment by melloware** [comment]
It's not prohibited but Monitor download only works with Ajax requests so it can monitor the start and stop. Non Ajax request is a full request response the "stop" won't ever be signaled. Which explains why you were seeing the dialog never go away 

And no Ajax attribute at all always defaults to TRUE for CommandButton. It's the default behavior.
